ntoskrnl/kdbg/amd64/kdb_help.S

```diff
@@ -1,22 +1,162 @@
 
 #include <asm.inc>
 
+#include <ksamd64.inc>
+
 .code64
 
+EXTERN KdbEnterDebuggerException : PROC
+
+PUBLIC KdbEnter
+FUNC KdbEnter
+
+    /* save flags */
+    pushfq
+    .allocstack 8
+
+    /* Make room for a KTRAP_FRAME */
+    sub rsp, KTRAP_FRAME_LENGTH
+    .allocstack KTRAP_FRAME_LENGTH
+    .ENDPROLOG
+
+    /* Save rbp */
+    mov [rsp + KTRAP_FRAME_Rbp], rbp
+
+    /* Save non-volatile registers */
+    mov [rsp + KTRAP_FRAME_Rbx], rbx
+    mov [rsp + KTRAP_FRAME_Rdi], rdi
+    mov [rsp + KTRAP_FRAME_Rsi], rsi
+
+    /* Save volatile registers */
+    mov [rsp + KTRAP_FRAME_Rax], rax
+    mov [rsp + KTRAP_FRAME_Rcx], rcx
+    mov [rsp + KTRAP_FRAME_Rdx], rdx
+    mov [rsp + KTRAP_FRAME_R8], r8
+    mov [rsp + KTRAP_FRAME_R9], r9
+    mov [rsp + KTRAP_FRAME_R10], r10
+    mov [rsp + KTRAP_FRAME_R11], r11
+
+    /* Save xmm registers */
+    movdqa [rsp + KTRAP_FRAME_Xmm0], xmm0
+    movdqa [rsp + KTRAP_FRAME_Xmm1], xmm1
+    movdqa [rsp + KTRAP_FRAME_Xmm2], xmm2
+    movdqa [rsp + KTRAP_FRAME_Xmm3], xmm3
+    movdqa [rsp + KTRAP_FRAME_Xmm4], xmm4
+    movdqa [rsp + KTRAP_FRAME_Xmm5], xmm5
+
+    /* Save cs and previous mode */
+    mov ax, cs
+    mov [rsp + KTRAP_FRAME_SegCs], ax
+    and ax, 1
```

Comment:
Can you explain how this is giving the previous mode?

Reply:
Grep for `KiUserTrap`

Reply:
sdk/include/asm/ksx.template.h	seems to have `CONSTANT(MODE_MASK)`, does this mean one could use such a define here in the ASM code? (If so, this would make the intent of the code more understandable.)

Reply:
It always gives KernelMode, like in the x86 case. It will in fact soon do nothing, because this function isn't even used anymore.